config/utility.py

```diff
@@ -13,3 +16,80 @@ def AddSystematicShift(
         producer[0].ignore_shift(name, producer[1])
     for producer in base_producers:
         producer[0].shift(name, producer[1])
+
+
+def ResolveSampleDependencies(config, sample):
+    for key in config.keys():
+        if isinstance(config[key], dict):
+            if list(config[key].keys())[0].startswith("SAMPLE_"):
+                if "SAMPLE_" + sample in config[key].keys():
+                    config[key] = config[key]["SAMPLE_" + sample]
+                else:
+                    log.error("Sample {} not available in {}!".format(sample, key))
+                    raise Exception
+            else:
+                ResolveSampleDependencies(config[key], sample)
+
+
+def ResolveEraDependencies(config, era):
+    for key in config.keys():
+        if isinstance(config[key], dict):
+            if list(config[key].keys())[0].startswith("ERA_"):
+                if "ERA_" + era in config[key].keys():
+                    config[key] = config[key]["ERA_" + era]
+                else:
+                    log.error("Era {} not available in {}!".format(era, key))
+                    raise Exception
+            else:
+                ResolveEraDependencies(config[key], era)
+
+
+def CollectProducerOutput(producer):
+    output = []
+    if producer.output != None:
+        if isinstance(producer.output, list):
+            output.extend(producer.output)
+        else:
+            output.append(producer.output)
+    if hasattr(producer, "producers"):
+        for prod in producer.producers:
+            output.extend(CollectProducerOutput(prod))
+    return output
+
+
+class ProducerRule:
+    def __init__(self, producers, samples, scopes, invert=False, update_output=True):
+        self.producers = producers
+        self.samples = samples
+        self.scopes = scopes
+        self.invert = invert
+        self.update_output = update_output
+
+    def applicable(self, sample):
+        applicable = sample in self.samples
+        if self.invert:
+            applicable = not applicable
+        return applicable
+
+    def operate(self, item, item_list):
```

Comment:
so this does nothing yet ?

Reply:
only in the derived classes

Reply:
only for doc that this should be defined in derived classes